source/v5/mqtt5_to_mqtt3_adapter.c

```diff
@@ -2200,16 +2206,15 @@ static void s_aws_mqtt5_to_mqtt3_adapter_subscribe_completion_fn(
 
         enum aws_mqtt_qos granted_qos = AWS_MQTT_QOS_AT_MOST_ONCE;
 
-        size_t subscription_count = aws_array_list_length(&subscribe_op->subscriptions);
         if (subscription_count > 0) {
             aws_array_list_get_at(&subscribe_op->subscriptions, &record, 0);
             topic_filter = record->subscription_view.topic_filter;
         }
 
-        if (suback != NULL) {
-            if (suback->reason_code_count > 0) {
-                granted_qos = s_convert_mqtt5_suback_reason_code_to_mqtt3_granted_qos(suback->reason_codes[0]);
-            }
+        if (reason_code_count > 0) {
+            granted_qos = s_convert_mqtt5_suback_reason_code_to_mqtt3_granted_qos(suback->reason_codes[0]);
+        } else if (record) {
```

Comment:
I am confused as to what's going on here.  Under no circumstances should we ever "invent" reason codes where none exist.  A suback with no reason codes in it (or a mismatch relative to the subscribe) should tear down the connection as a protocol error.  If we're not doing so atm, we should.

Reply:
MQTT3 on a timeout will return a timeout error code with the original subscribe request topic and request qos. If we have no suback packet due to a timeout or shutting down the client in mqtt5, to emulate mqtt3 behavior we need to also send back the same along with the error code.

Reply:
haha, yea. I had the same thought when I checked. The sins of our fathers...

Reply:
Do any bindings do anything with those values when an error is set?  There are (buggy) cases where an error may not get set properly in which case doing this makes it look like the subscribe succeeded.

Reply:
Yep, aws-crt-cpp extracts info regardless of the error code: https://github.com/awslabs/aws-crt-cpp/blob/main/source/mqtt/MqttConnectionCore.cpp#L417

Reply:
Python bindings returns a future on a subscribe() and if there's an error code, sets an exception from the error_code in the future...

Reply:
Yikes. So in cpp, it's up to the customer to check the error code to make sure they're not assuming successful subscribes. I need to double-check that we're not at least setting the request.qos to fail in a timeout. If we are, then we should be fine everywhere

Reply:
Java binding will directly return error_code, while JS binding will extract the data regardless of the error_code. 

It is probably reasonable to keep the topic and qos data when the subscription failed with an error_code. The user still wanna get the information of a failing subscription for possible operations. (For a simply example, the user would need the info to print out a proper error message, or try subscribe again on the same topic/qos value.)